File: Medium/3066. Minimum Operations to Exceed Threshold Value II/minOperations.cpp
#pragma GCC optimize("O3", "unroll-loops")
#pragma GCC target("avx,mmx,sse2,sse3,sse4")

class Solution {
public:
    int minOperations(vector<int>& nums, int k) {
        u_int x = 0, y = 0, size, index, rightChild, smallest;
        size = nums.size();
        for (; y < size; ++y) {
            if (nums[y] < k) {
                nums[x++] = nums[y];
            }
        }
        nums.resize(x);
        size = x;
        make_heap(nums.begin(), nums.end(), greater<int>());
        x = 0;
        while (size > 1) {

            y = nums.front() * 2;
            pop_heap(nums.begin(), nums.end(), greater<int>());
            nums.pop_back();

            y += nums.front();

            if (y < k) {
                nums.front() = y;
                size--;
                index = 0;
                while (true) {
                    rightChild = 2 * index + 2;
                    smallest = rightChild - 1;

                    if (rightChild < size &&
                        nums[rightChild] < nums[smallest]) {
                        smallest = rightChild;
                    }

                    if (smallest >= size || nums[smallest] >= y) {
                        break;
                    }

                    nums[index] = nums[smallest];
                    index = smallest;
                }

                nums[index] = y;
            } else {
                pop_heap(nums.begin(), nums.end(), greater<int>());
                nums.pop_back();
                size -= 2;
            }
            x++;
        }
        if (nums.size() == 1) {
            x++;
        }
        return x;
    }
};

File: Medium/3066. Minimum Operations to Exceed Threshold Value II/minOperations.py
import heapq


class Solution:
    def minOperations(self, nums: List[int], k: int) -> int:
        x = len(nums)
        y = 0
        while y < x:
            if nums[y] >= k:
                x -= 1
                while nums[x] >= k and x > y:
                    x -= 1
                nums[y] = nums[x]
            y += 1
        del nums[x:]

        heapq.heapify(nums)
        x = 0
        while len(nums) > 1:
            y = heapq.heappop(nums) * 2 + nums[0]
            if y < k:
                heapq.heapreplace(nums, y)
            else:
                heapq.heappop(nums)
            x += 1

        return x + 1 if nums else x
